providers/implementations/ciphers/cipher_aes_gcm_siv_hw.c

```diff
@@ -0,0 +1,373 @@
+        len_blk[1] = BSWAP8((uint64_t)len * 8);
+    }
+    memset(S_s, 0, TAG_SIZE);
+    ossl_polyval_ghash_init(ctx->Htable, (const uint64_t*)ctx->msg_auth_key);
+    if (ctx->aad != NULL) {
+        /* AAD allocated with padding, but need to adjust length */
+        ossl_polyval_ghash_hash(ctx->Htable, S_s, ctx->aad, UP16(ctx->aad_len));
+    }
+    if (DOWN16(len) > 0)
+        ossl_polyval_ghash_hash(ctx->Htable, S_s, out, DOWN16(len));
+    if (!IS16(len)) {
+        /* deal with padding - probably easier to "memset" the padding first rather than calculate */
+        padding[0] = padding[1] = 0;
+        memcpy(padding, &out[DOWN16(len)], REMAINDER16(len));
+        ossl_polyval_ghash_hash(ctx->Htable, S_s, (uint8_t *)padding, sizeof(padding));
+    }
+    ossl_polyval_ghash_hash(ctx->Htable, S_s, (uint8_t *)len_blk, TAG_SIZE);
+
+    for (i = 0; i < NONCE_SIZE; i++)
+        S_s[i] ^= ctx->nonce[i];
+
+    S_s[TAG_SIZE - 1] &= 0x7f;
+
+    /*
+     * In the ctx, user_tag is the one received/set by the user,
+     * and tag is generated from the input
+     */
+    out_len = sizeof(ctx->tag);
+    error |= !EVP_EncryptUpdate(ctx->ecb_ctx, ctx->tag, &out_len, S_s, sizeof(S_s));
+    ctx->generated_tag = !error;
+    /* Regardless of error */
+    ctx->used_dec = 1;
+    return !error;
+}
+
+static int aes_gcm_siv_cipher(void *vctx, unsigned char *out,
+                              const unsigned char *in, size_t len)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+
+    /* EncryptFinal or DecryptFinal */
+    if (in == NULL)
+        return aes_gcm_siv_finish(ctx);
+
+    /* Deal with associated data */
+    if (out == NULL)
+        return aes_gcm_siv_aad(ctx, in, len);
+
+    if (ctx->enc)
+        return aes_gcm_siv_encrypt(ctx, in, out, len);
+
+    return aes_gcm_siv_decrypt(ctx, in, out, len);
+}
+
+static void aes_gcm_siv_clean_ctx(void *vctx)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+
+    EVP_CIPHER_CTX_free(ctx->ecb_ctx);
+    ctx->ecb_ctx = NULL;
+}
+
+static int aes_gcm_siv_dup_ctx(void *vdst, void *vsrc)
+{
+    PROV_AES_GCM_SIV_CTX *dst = (PROV_AES_GCM_SIV_CTX *)vdst;
+    PROV_AES_GCM_SIV_CTX *src = (PROV_AES_GCM_SIV_CTX *)vsrc;
+
+    dst->ecb_ctx = NULL;
+    if (src->ecb_ctx != NULL) {
+        if ((dst->ecb_ctx = EVP_CIPHER_CTX_new()) == NULL)
+            goto err;
+        if (!EVP_CIPHER_CTX_copy(dst->ecb_ctx, src->ecb_ctx))
+            goto err;
+    }
+    return 1;
+
+ err:
+    EVP_CIPHER_CTX_free(dst->ecb_ctx);
+    dst->ecb_ctx = NULL;
+    return 0;
+}
+
+static const PROV_CIPHER_HW_AES_GCM_SIV aes_gcm_siv_hw =
+{
+    aes_gcm_siv_initkey,
+    aes_gcm_siv_cipher,
+    aes_gcm_siv_dup_ctx,
+    aes_gcm_siv_clean_ctx,
+};
+
+const PROV_CIPHER_HW_AES_GCM_SIV *ossl_prov_cipher_hw_aes_gcm_siv(size_t keybits)
+{
+    return &aes_gcm_siv_hw;
+}
+
+/* NEED AES-CTR32... which is a 32bit counter version of AES-CTR which is 128? */
```

Comment:
nit: We probably shouldn't have uncertain comments like this in the code. The RFC specifies a 32-bit counter, so there's no ambiguity here

Reply:
That's fair. The comment was really really related to the definition of AES-CTR using a 128bit counter.